pkg/controller/connectivitycheck/connectivity_check_controller.go

```diff
@@ -0,0 +1,404 @@
+		return nil
+	}
+
+	apiUrl, err := url.Parse(infrastructure.Status.APIServerURL)
+	if err != nil {
+		recorder.Warningf("EndpointDetectionFailure", "error detecting external api load balancer endpoint: %v", err)
+
+	} else {
+		templates = append(templates, NewPodNetworkConnectivityCheckTemplate(apiUrl.Host, "openshift-network-diagnostics", withTarget("load-balancer", "api-external")))
+	}
+
+	apiInternalUrl, err := url.Parse(infrastructure.Status.APIServerInternalURL)
+	if err != nil {
+		recorder.Warningf("EndpointDetectionFailure", "error detecting internal api load balancer endpoint: %v", err)
+	} else {
+		templates = append(templates, NewPodNetworkConnectivityCheckTemplate(apiInternalUrl.Host, "openshift-network-diagnostics", withTarget("load-balancer", "api-internal")))
+	}
+	return templates
+}
+
+type endpointInfo struct {
+	hostName string
+	port     string
+	nodeName string
+}
+
+func withTarget(label, target string) func(check *v1alpha1.PodNetworkConnectivityCheck) {
+	return WithTarget(label + "-" + target)
+}
+
+func Start(ctx context.Context, kubeConfig *rest.Config) error {
+	protoKubeConfig := rest.CopyConfig(kubeConfig)
+	protoKubeConfig.ContentType = "application/vnd.kubernetes.protobuf,application/json"
+	eventRecorder := &eventrecorder.LoggingRecorder{}
+	kubeClient, err := kubernetes.NewForConfig(protoKubeConfig)
+	if err != nil {
+		return err
+	}
+	configClient, err := configv1client.NewForConfig(kubeConfig)
+	if err != nil {
+		return err
+	}
+	operatorClient, dynamicInformers, err := genericoperatorclient.NewClusterScopedOperatorClient(kubeConfig, operatorv1.GroupVersion.WithResource("openshiftapiservers"))
+	if err != nil {
+		return err
+	}
+	operatorcontrolplaneClient, err := operatorcontrolplaneclient.NewForConfig(kubeConfig)
+	if err != nil {
+		return err
+	}
+	apiextensionsClient, err := apiextensionsclient.NewForConfig(kubeConfig)
+	if err != nil {
+		return err
+	}
+	apiextensionsInformers := apiextensionsinformers.NewSharedInformerFactory(apiextensionsClient, 10*time.Minute)
+	kubeInformersForNamespaces := v1helpers.NewKubeInformersForNamespaces(kubeClient,
+		"",
+		"openshift-config",
+		"openshift-config-managed",
+		"openshift-cluster-network-operator",
+		"openshift-network-diagnostics",
+		metav1.NamespaceSystem,
+		"openshift-kube-apiserver",
+		"openshift-apiserver",
+		"openshift-oauth-apiserver",
+	)
+	configInformers := configinformers.NewSharedInformerFactory(configClient, 10*time.Minute)
+	connectivityCheckController := NewNetworkConnectivityCheckController(
+		kubeClient,
+		operatorClient,
+		operatorcontrolplaneClient,
+		apiextensionsClient,
+		kubeInformersForNamespaces,
+		configInformers,
+		apiextensionsInformers,
+		eventRecorder,
+	)
+
+	go connectivityCheckController.Run(ctx, 1)
+	apiextensionsInformers.Start(ctx.Done())
+	kubeInformersForNamespaces.Start(ctx.Done())
+	dynamicInformers.Start(ctx.Done())
+	configInformers.Start(ctx.Done())
```

Comment:
Shouldn't you start the informers before running the controller?

Reply:
and wait for cache sync? But if you switch to `pkg/client` then that's taken care for you.

Reply:
done